### PR TITLE
sidecar: trim service ports based on egress listeners and merge services

### PR DESCRIPTION
**Pull request by ramaraochavali**
Reverts `Revert https://github.com/istio/istio/pull/18212 of https://github.com/istio/istio/pull/18111.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@rshriram Found a minor issue with the code with the config used in e2e tests. I pushed with out my fix and it failed. With fix it passed consistently four times. I have added additional test to cover that.

@howardjohn PTAL, FYI. Sorry for the noise

File: pilot/pkg/model/sidecar.go [hunk 1]
```diff
@@ -220,11 +218,25 @@ func ConvertToSidecarScope(ps *PushContext, sidecarConfig *Config, configNamespa
...
-			// TODO: port merging when each listener generates a partial service
 			if _, found := servicesAdded[string(s.Hostname)]; !found {
-				servicesAdded[string(s.Hostname)] = struct{}{}
+				servicesAdded[string(s.Hostname)] = s
 				out.services = append(out.services, s)
 				out.namespaceDependencies[s.Attributes.Namespace] = struct{}{}
+			} else if s.Ports != nil && len(s.Ports) > 0 {
+				// merge the ports to service when each listener generates partial service
+				os := servicesAdded[string(s.Hostname)]
+				for _, p := range s.Ports {
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar.go]: @rshriram This is the fix I have put. Earlier, it was duplicate ports while merging.

File: pilot/pkg/model/sidecar_test.go [hunk 2]
```diff
@@ -107,6 +146,78 @@ var (
...
+						Protocol: "HTTP",
+						Name:     "uds",
+					},
+					Hosts: []string{"foo/*"},
+				},
+			},
+		},
+	}
+
+	configs7 = &Config{
```
Comment (ramaraochavali) [inline pilot/pkg/model/sidecar_test.go]: This is new config test matching the e2e test

File: pilot/pkg/model/sidecar.go (hunk 1 above)
Comment (rshriram) [reply pilot/pkg/model/sidecar.go]: How is duplicate ports possible in an egress listener though?

Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go]: good point. I did see that in debug when services are merged by listener, each listener has the same service with same port - Now that you mentioned this point, seems there is a problem. Let me debug again.

File: pilot/pkg/model/service.go [hunk 3]
```diff
@@ -653,3 +654,37 @@ func (s *Service) GetServiceAddressForProxy(node *Proxy) string {
...
+		CreationTime:    s.CreationTime,
+		Hostname:        s.Hostname,
+		Address:         s.Address,
+		ClusterVIPs:     clusterVIPs.(map[string]string),
+		Resolution:      s.Resolution,
+		MeshExternal:    s.MeshExternal,
+	}
+}
+
+func copyInternal(v interface{}) interface{} {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/service.go]: Is there a benchmark of `copystructure`? Is that more efficient compared to copy one by one?

Comment (ramaraochavali) [reply pilot/pkg/model/service.go]: I do not know of any benchmark of copy structure. I

Comment (ramaraochavali) [reply pilot/pkg/model/service.go]: Do you prefer not using copystructure and just copy every thing ourselves?

File: pilot/pkg/model/sidecar.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/model/sidecar.go]: If you are wondering about test case, I incorrectly changed the test case earlier - which I corrected in this PR as well

**Comment by ramaraochavali** [comment]
I ran `/test pilot-multicluster-e2e_istio` test multiple times today and it passed all the time except in two cases for unrelated reasons. So @rshriram if you are good with the code changes, I will rerun the unit test (did not do because - it might merge).

File: pilot/pkg/model/service.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/service.go]: I prefer the more efficient way.

Comment (ramaraochavali) [reply pilot/pkg/model/service.go]: Ok. I will go with this now and run a test and see if it is better to copy individuals. I will do a follow-up if it turns out individual copy is better

**Pull request merged**